Add tests to language part 5

bingbing8 opened the pull request:
Add parameter binding tests.

msftclas commented:
Hi __@bingbing8__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Yanbing Wang). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

vors commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -0,0 +1,568 @@
+        }
+    }   
+    
+    #know issue
```
known

bingbing8 replied:
fixed.

---
In reply to: [78272031](https://github.com/Powershell/Powershell/pull/2218#discussion_r78272031) [](ancestors = 78272031)

lzybkr commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -0,0 +1,577 @@
+    It '<name>' -TestCases:$testData{
+        param($name, $cmd, $expected)
+        
+        $actual = Invoke-Expression $cmd
```
I see no reason for `Invoke-Expression`.  I think using script blocks makes more sense.

bingbing8 replied:
fixed.

---
In reply to: [78435988](https://github.com/Powershell/Powershell/pull/2218#discussion_r78435988) [](ancestors = 78435988)

lzybkr commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -0,0 +1,577 @@
+        }
+    }
+
+    $testData = @(
```
I'm not sure this is a good use of `-TestCases`. It essentially forces a separation of the test function and the validation making it harder to really see what's being tested.  Many of the functions are only called once, so I think it doesn't make sense to use `-TestCases` for those at all. The others, maybe it makes sense, it's hard to say for sure though.

bingbing8 replied:
Thanks. I will talk to Jim before i make changes. 

---
In reply to: [78436715](https://github.com/Powershell/Powershell/pull/2218#discussion_r78436715) [](ancestors = 78436715)

bingbing8 replied:
I made the below changes. Please review:
1. For those functions being reused by different variations, I call Validate-Context where the function is passed as Setup parameter and is called in context and the validation is done in the It block which takes the variation as test cases.
2. For those functions being used by only one tests, we pass the function as Setup item and is called in the It block.


---
In reply to: [78436715](https://github.com/Powershell/Powershell/pull/2218#discussion_r78436715) [](ancestors = 78436715)

bingbing8 committed:
1. Added parameterbinding test
2. Added ShouldBeErrorID function in helper file
3. Removed ShouldBeErrorID function from other test modules
4. Update New-TestHost to be able to run on full CLR
5. updated file map.json

JamesWTruher commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -0,0 +1,538 @@
+    It 'Multiple positional parameters case 2' {
+        function get-foo
+        {
+            param(
```
>param [](start = 12, length = 5)

could you add a comment here indicating that the parameters are purposefully out of order?

bingbing8 replied:
Sure, added in next iteration

---
In reply to: [79213263](https://github.com/Powershell/Powershell/pull/2218#discussion_r79213263) [](ancestors = 79213263)

JamesWTruher commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -0,0 +1,538 @@
+
+    }
+
+    Context 'default values not converted properly' {
```
>default values not converted properly [](start = 13, length = 37)

these titles need to be changed to reflect the the positive test, eg:
`Context 'Default value conversion tests'`
and
`It 'Parameter default value is converted correctly to the proper type'`
it looks like a number of these should be changed

bingbing8 replied:
fixed in next iteration

---
In reply to: [79214652](https://github.com/Powershell/Powershell/pull/2218#discussion_r79214652) [](ancestors = 79214652)

bingbing8 committed:
Updated the titles

Updated the titles of context and it to be more describle.

lzybkr commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -0,0 +1,539 @@
+    }
+
+    It 'Multiple positional parameters case 1' {
+       
```
I feel like there is a lot of extra white space in this block (and elsewhere) - increasing the code density makes it easier to read in my opinion.  I'd remove some of the blank lines, and maybe even make the `param` blocks a bit more compact, e.g. not putting the closing `)` on it's own line.

bingbing8 replied:
fixed in next iteration

---
In reply to: [79250423](https://github.com/Powershell/Powershell/pull/2218#discussion_r79250423) [](ancestors = 79250423)

lzybkr commented on test/powershell/Language/Scripting/ParameterBinding.Tests.ps1:
```diff
@@ -0,0 +1,539 @@
+        { get-foo -a a -b b c d } | ShouldBeErrorId 'AmbiguousParameterSet,get-foo'
+        
+
+        ( get-foo -a a b c d ) -join ',' | Should Be 'b,c,d'
```
I wonder if it's cleaner/better to add a filter to do the join, e.g.
```powershell
get-foo -a a b c d | Join ',' | Should Be 'b,c,d'
```
Pros - it feels more like PowerShell
Cons - it would be slower

bingbing8 replied:
I decided not to implement new function.

---
In reply to: [79251051](https://github.com/Powershell/Powershell/pull/2218#discussion_r79251051) [](ancestors = 79251051)

lzybkr reviewed (changes requested):
These changes are looking good, it was much easier to review.  I'd just like some less white space and then it looks good.

bingbing8 commented on test/powershell/Language/Parser/UsingNamespace.Tests.ps1:
```diff
@@ -10,27 +10,7 @@ using namespace System.Diagnostics; using namespace System.Runtime.CompilerServi
-        $_.FullyQualifiedErrorId | Should Be $FullyQualifiedErrorId
-    }
-}
+Import-Module -Force $PSScriptRoot\..\..\Common\Test.Helpers.psm1
```
When develop tests, yes. If there is changes in the module, the test does not pick up the latest changes without force. but it does not matter during test run.


---
In reply to: [79249686](https://github.com/Powershell/Powershell/pull/2218#discussion_r79249686) [](ancestors = 79249686)

bingbing8 replied:
I will removed the Force then

---
In reply to: [79257879](https://github.com/Powershell/Powershell/pull/2218#discussion_r79257879) [](ancestors = 79257879,79249686)

bingbing8 committed:
Remove some blank lines and white spaces

Remove some blank lines and white spaces to make the code more compact.

bingbing8 committed:
Removed the usage of Force

Removed the usage of Force

bingbing8 commented:
@lzybkr I removed a lot of white spaces and blank lines. Please review. Thanks!

vors commented:
@lzybkr @bingbing8 7 tests failed in master after this PR was merged.

lzybkr commented:
The failing tests are in PackageManagement, this PR added unrelated tests, so it's most likely an external factor that caused those tests to fail.

vors commented:
I will re-trigger the build on master

vors commented:
Actually, nevermind, it was a daily build. So the failures are expected.
